main.go

```diff
@@ -0,0 +1,323 @@
+	//http://stackoverflow.com/a/4073451/4710047
+
+	//expecting to come from PUT data field
+	taskName := r.PostFormValue("name")
+
+	//start transaction
+	tx, err := db.Begin()
+	if err != nil {
+		log.Fatal(err)
+	}
+	// insert query
+	stmt, err := tx.Prepare("insert into Task(name) values(?)")
+	if err != nil {
+		log.Fatal(err)
+	}
+	//close statement
+	defer stmt.Close()
+	//execute statement
+	resp, err := stmt.Exec(taskName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	//commit statement in db
+	tx.Commit()
+
+	//get last inserted task id in order to show it to the user
+	newTaskId, err := resp.LastInsertId()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	//create new task model
+	newTask := models.Task{Id: int(newTaskId), Name: taskName}
+
+	//encode newTask as json, and return it to user.
+	if err := json.NewEncoder(w).Encode(newTask); err != nil {
+		panic(err)
+	}
+}
+
+func UpdateTask(w http.ResponseWriter, r *http.Request) {
+	//note: r.FormValue searches for key in POST data fields, then PUT data fields
+	//2 types of POST submissions: application/x-www-form-urlencoded AND multipart/form-data.
+	// need to understand both. Generally speaking, urlencoded takes up extra space so is for normal post requests. multipart form-data does not increase space usage by a lot so is for uploading files
+	//http://stackoverflow.com/a/4073451/4710047
+
+	log.Printf("Update Task")
+	//create a variable that has the parameters sent to the api in the url
+	// /task/<id>/<x>/  will lead to variables id and x in vars
+	vars := mux.Vars(r)
+
+	//get task id
+	taskId, err := strconv.Atoi(vars["id"])
+	if err != nil {
+		log.Fatal(err)
+	}
+	//get new name from POST data form
+	newName := r.PostFormValue("name")
+
+	//start the transaction (do not have to close transaction, but do have to commit them)
+	tx, err := db.Begin()
+	if err != nil {
+		log.Fatal(err)
+	}
+	//update query statement.
+	stmt, err := tx.Prepare("UPDATE Task SET  name=? WHERE id=?")
+	if err != nil {
+		log.Fatal(err)
+	}
+	//close statement
+	defer stmt.Close()
+	//check for errors. response does not give you the row that had the update so _
+	_, err = stmt.Exec(newName, taskId)
+	if err != nil {
+		log.Fatal(err)
+	}
+	//commit the transaction
+	tx.Commit()
+
+	//create model for updated task
+	updatedTask := models.Task{Id: int(taskId), Name: newName}
+
+	//encode task as json and return to user
+	if err := json.NewEncoder(w).Encode(updatedTask); err != nil {
+		panic(err)
+	}
+}
+
+//view a single task
+//input: id of task as url parameter
+//output: json serialization of task with input id
+func ViewTask(w http.ResponseWriter, r *http.Request) {
+
+	//note: r.FormValue searches for key in GET queries, then POST data fields, then PUT data fields
+	log.Printf("View Task")
+	//create vars variable to access the url parameters
+	vars := mux.Vars(r)
+
+	//get id from url params
+	taskId, err := strconv.Atoi(vars["id"])
```

Comment:
put in a comment saying that vars[id] will always exist else this function will not be called.